text/0000-offset_of.md

```diff
@@ -0,0 +1,482 @@
+```
+
+Finally, `offset_of!` can be used to compute the offset of fields in unions too.
+While this may be surprising, the compiler is allowed to put padding in front of
+fields in unions which are not `#[repr(C)]`, which would lead to a non-zero
+field offset.
+
+```rs
+use core::mem::offset_of;
+union Buffer {
+    metadata: [u64; 3],
+    datadata: [u8; 1024 * 1024 * 32],
+}
+const METADATA_OFFSET: usize = offset_of!(Buffer, metadata);
+```
+
+## Limitations
+
+There are a few limitations worth mentioning. Some of these may be relaxed in
+the future, however.
+
+1. Perhaps unsurprisingly, it obeys privacy, so both the type and field you call
+   `offset_of!` on must be visible to the code calling `offset_of!`.
+
+2. The type holding the field must be `Sized`, so trying to compute where the
+   slice begins in something like `offset_of!((i32, [u32]), 1)` isn't supported.
+
+3. Compared to `offsetof` in C and C++, you can't access nested fields/arrays.
+   That is, instead of `offset_of!(Foo, quank.zoop.2.quank[4])`, you'll have to
+   compute the offsets of each step manually, and sum them.
+
+4. Finally, types other than tuples, structs, and unions are currently
+   unsupported.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`offset_of` is a new macro exported from `core::mem` which has a signature
+similar to the following:
+
+```rs
+pub macro offset_of($Container:ty, $field:tt $(,)?) {
+    // ...implementation defined...
+}
+```
+
+Invoking this macro expands to a constant expression of type `usize`, which
+evaluates to the offset in bytes from the beginning of `$Container` where
+`$field` is found.
+
+`$Container` must be visible and must be or resolve to one of the following
+types:
+
+1. A `struct` or `union` type with either named or anonymous/tuple-style fields.
+
+    In this case, `$field` must share a name or tuple index with a field which:
+    - Exists on `$Container`.
+    - Is visible at the location where `offset_of!` is invoked (but there is no
+      requirement that fields other than than `$field` be visible there)
+
+2. An anonymous tuple type.
+
+    In this case, `$field` must be a tuple index (that is, an integer literal)
+    that exists on the tuple type in question.
+
+Use on other types is an error, although this may be relaxed in some cases in
+the future (see the [Future possibilities][future-possibilities] section).
+
+As a note: the implementation is strongly encouraged to not have runtime
+resource usage dependent on the values of `$Container` or `$field`. In
+particular, the implementation should not allocate space for an instance of
+`$Container` on the runtime stack.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. This exposes layout information at compile time which is otherwise not
+   exposed until runtime. This can cause compatibility hazards similar to
+   `mem::size_of` or `mem::align_of`, but plausibly greater as it provides even
+   more information.
```

Comment:
Note that with rust-lang/rust#96240 (`const_ptr_offset_from`) being merged for 1.65, this will already be available in a const context - it was the only feature gate left for `memoffset::offset_of!()` to be able to be const (as long as the type doesn't have UnsafeCell somewhere, that's `const_refs_to_cell`).